core/Pos/Binary/Class/Core.hs

```diff
@@ -1,171 +1,589 @@
-
--- | Find out how many bytes were read during a 'Peek'.
-getPeekLength :: Peek a -> Peek (a, Int)
-getPeekLength m = Peek $ \ps ptr -> do
-    PeekResult ptr2 x <- Store.runPeek m ps ptr
-    return (PeekResult ptr2 (x, ptr2 `minusPtr` ptr))
-{-# INLINE getPeekLength #-}
+instance Bi Millisecond where
+    encode = encode . toInteger
+    decode = fromInteger <$> decode
+
+instance Bi Microsecond where
+    encode = encode . toInteger
+    decode = fromInteger <$> decode
+
+instance Bi Byte where
+    encode = encode . toBytes
+    decode = fromBytes <$> decode
+
+genericEncode :: (Generic a, GSerialiseEncode (G.Rep a)) => a -> E.Encoding
+genericEncode = gencode . G.from
+
+genericDecode :: (Generic a, GSerialiseDecode (G.Rep a)) => D.Decoder s a
+genericDecode = G.to <$> gdecode
+
+class GSerialiseEncode f where
+    gencode  :: f a -> E.Encoding
+
+class GSerialiseDecode f where
+    gdecode  :: D.Decoder s (f a)
+
+instance GSerialiseEncode G.V1 where
+    -- Data types without constructors are still serialised as null value
+    gencode _ = E.encodeNull
+
+instance GSerialiseDecode G.V1 where
+    gdecode   = error "G.V1 don't have contructors" <$ D.decodeNull
+
+instance GSerialiseEncode G.U1 where
+    -- Constructors without fields are serialised as null value
+    gencode _ = E.encodeListLen 1 <> E.encodeWord 0
+
+instance GSerialiseDecode G.U1 where
+    gdecode   = do
+      n <- D.decodeListLen
+      when (n /= 1) $ fail "expect list of length 1"
+      tag <- D.decodeWord
+      when (tag /= 0) $ fail "unexpected tag. Expect 0"
+      return G.U1
+
+instance GSerialiseEncode a => GSerialiseEncode (G.M1 i c a) where
+    -- Metadata (constructor name, etc) is skipped
+    gencode = gencode . G.unM1
+
+instance GSerialiseDecode a => GSerialiseDecode (G.M1 i c a) where
+    gdecode = G.M1 <$> gdecode
+
+instance Bi a => GSerialiseEncode (G.K1 i a) where
+    -- Constructor field (Could only appear in one-field & one-constructor
+    -- data types). In all other cases we go through GSerialise{Sum,Prod}
+    gencode (G.K1 a) = E.encodeListLen 2
+                  <> E.encodeWord 0
+                  <> encode a
+
+instance Bi a => GSerialiseDecode (G.K1 i a) where
+    gdecode = do
+      n <- D.decodeListLen
+      when (n /= 2) $
+        fail "expect list of length 2"
+      tag <- D.decodeWord
+      when (tag /= 0) $
+        fail "unexpected tag. Expects 0"
+      G.K1 <$> decode
+
+instance (GSerialiseProd f, GSerialiseProd g) => GSerialiseEncode (f G.:*: g) where
+    -- Products are serialised as N-tuples with 0 constructor tag
+    gencode (f G.:*: g)
+        = E.encodeListLen (nFields (Proxy :: Proxy (f G.:*: g)) + 1)
+       <> E.encodeWord 0
+       <> encodeSeq f
+       <> encodeSeq g
+
+instance (GSerialiseProd f, GSerialiseProd g) => GSerialiseDecode (f G.:*: g) where
+    gdecode = do
+      let nF = nFields (Proxy :: Proxy (f G.:*: g))
+      n <- D.decodeListLen
+      -- TODO FIXME: signedness of list length
```

Comment:
I should have clarified all the `Generic` code you see here has been copied in bulk from WT's `serialise` library, so we should have our back covered. Something I will mention here so I don't forget is that @dcoutts mentioned how we should at a later stage optimise this to remove the redundant "version" tag the `serialise` library was adding to single constructor types. This is needed in `serialise` to deal with "schema changes" and migrations, but it's something we don't need here.